syntax/syntax.go

```diff
@@ -9,6 +9,47 @@ package syntax
 type Node interface {
 	// Span returns the start and end position of the expression.
 	Span() (start, end Position)
+
+	// Comments returns the comments associated with this node.
+	// It returns nil if RetainComments was not specified during parsing.
+	Comments() *Comments
+
+	// AllocComments allocates a new Comments node if there was none.
+	// This makes possible to add new comments using Comments() method.
+	AllocComments()
+}
+
+// A Comment represents a single # comment.
+type Comment struct {
+	Start  Position
+	Text   string // without trailing newline
+	Suffix bool   // an end-of-line (not whole line) comment
+}
+
+// Comments collects the comments associated with an expression.
+type Comments struct {
+	Before []Comment // whole-line comments before this expression
+	Suffix []Comment // end-of-line comments after this expression
+
+	// For top-level expressions only, After lists whole-line
+	// comments following the expression.
+	After []Comment
+}
+
+// A CommentsRef is a possibly-nil reference to a set of comments.
```

Comment:
type commentsRef (lowercase)

Reply:
This breaks the tests:

--- FAIL: TestExprParseTrees (0.00s)
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
	panic: reflect.Value.Interface: cannot return value obtained from unexported field or method

Reply:
The loop over struct fields in treeString will need to skip the commentRef field.